File: Heap/javascript/heap_insert.js
function BinaryHeap(){
    let list = [];
    

    this.maxHeapify = (arr, n, i) => {
        let largest = i;
        let l = 2 * i + 1;
        let r = 2 * i + 2;

        if (l < n && arr[l] > arr[largest]) {
            largest = l; 
        }

        if (r < n && arr[r] > arr[largest]) {
            largest = r; 
        }

        if (largest != i) { 
            let temp = arr[i]; 
            arr[i] = arr[largest]; 
            arr[largest] = temp; 
            this.maxHeapify(arr, n, largest); 
        } 
    }
    

    this.insert = (num) => {
        const size = list.length;

        if(size === 0){
            list.push(num);
        }else{
            list.push(num);
            for (let i = parseInt(list.length / 2 - 1); i >= 0; i--) {
                this.maxHeapify(list, list.length, i); 
            }
        }
    }
    
    this.getList = () => list;
}



var heap = new BinaryHeap();
heap.insert(3);
heap.insert(4);
heap.insert(9);
heap.insert(5);
heap.insert(2);
heap.insert(1);

console.log("Max-Heap: ");
console.log(heap.getList());

File: Heap/java/heap_insert.java


import java.util.ArrayList;

public class heap_insert {
    void heapify(ArrayList<Integer> hT, int i) {
        int size = hT.size();
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        if (l < size && hT.get(l) > hT.get(largest))
        largest = l;
        if (r < size && hT.get(r) > hT.get(largest))
            largest = r;

        if (largest != i) {
            int temp = hT.get(largest);
            hT.set(largest, hT.get(i));
            hT.set(i, temp);
            heapify(hT, largest);
        }
    }

    void insert(ArrayList<Integer> hT, int newNum) {
        int size = hT.size();
        if (size == 0) {
            hT.add(newNum);
        } else {
            hT.add(newNum);
            for (int i = size / 2 - 1; i >= 0; i--) {
                heapify(hT, i);
            }
        }
    }


    void printArray(ArrayList<Integer> array, int size) {
        for (Integer i : array) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

    public static void main(String args[]) {

        ArrayList<Integer> array = new ArrayList<Integer>();
        int size = array.size();

        heap_insert h = new heap_insert();
        h.insert(array, 3);
        h.insert(array, 4);
        h.insert(array, 9);
        h.insert(array, 5);
        h.insert(array, 2);
        h.insert(array, 1);

        System.out.println("Max-Heap array: ");
        h.printArray(array, size);

    }

}

